internal/grpcsync/pubsub_test.go

```diff
@@ -0,0 +1,201 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package grpcsync
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+type testSubscriber struct {
+	mu      sync.Mutex
+	msgs    []int
+	onMsgCh chan struct{}
+}
+
+func newTestSubscriber() *testSubscriber {
+	return &testSubscriber{onMsgCh: make(chan struct{})}
+}
+
+func (ts *testSubscriber) OnMessage(msg interface{}) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	ts.msgs = append(ts.msgs, msg.(int))
+	ts.onMsgCh <- struct{}{}
+}
+
+func (ts *testSubscriber) receivedMsgs() []int {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+	return ts.msgs
+}
+
+func (s) TestPubSub_PublishNoMsg(t *testing.T) {
+	// Create a new pubsub.
+	pubsub := NewPubSub()
+	defer pubsub.Stop()
+
+	ts := newTestSubscriber()
+	pubsub.Subscribe(ts)
+
+	select {
+	case <-ts.onMsgCh:
+		t.Fatalf("Subscriber callback invoked when no message was published")
+	case <-time.After(defaultTestShortTimeout):
+	}
+}
+
+func (s) TestPutSub_PublishMsgs_RegisterSubs_And_Stop(t *testing.T) {
+	// This flag becomes true when waiting callback is timed out.
+	isTimeout := false
+	// Create a new pubsub.
+	pubsub := NewPubSub()
+
+	ts := newTestSubscriber()
+
+	pubsub.Subscribe(ts)
+	wantMsgs := []int{}
+
+	const numPublished = 10
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			pubsub.Publish(i)
+		}
+		wg.Done()
+	}()
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			select {
+			case <-ts.onMsgCh:
+				wantMsgs = append(wantMsgs, i)
+			case <-time.After(defaultTestShortTimeout):
+				isTimeout = true
+			}
+		}
+		wg.Done()
+	}()
+	wg.Wait()
+	if isTimeout {
+		t.Fatalf("Timeout when expecting the onMessage() callback to be invoked")
+	}
+	if gotMsgs := ts.receivedMsgs(); !cmp.Equal(gotMsgs, wantMsgs) {
+		t.Fatalf("Received messages is %v, want %v", gotMsgs, wantMsgs)
+	}
+
+	ts2 := newTestSubscriber()
+	pubsub.Subscribe(ts2)
+	wantMsgs2 := []int{}
+
+	select {
+	case <-ts2.onMsgCh:
+		wantMsgs2 = append(wantMsgs2, numPublished-1)
+	case <-time.After(defaultTestShortTimeout):
+		t.Fatalf("Timeout when expecting the onMessage() callback to be invoked")
+	}
+	if gotMsgs2 := ts2.receivedMsgs(); !cmp.Equal(gotMsgs2, wantMsgs2) {
+		t.Fatalf("Received messages is %v, want %v", gotMsgs2, wantMsgs2)
+	}
+
+	var wg2 sync.WaitGroup
+	wg2.Add(3)
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			pubsub.Publish(i)
+		}
+		wg2.Done()
+	}()
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			select {
+			case <-ts.onMsgCh:
+				wantMsgs = append(wantMsgs, i)
+			case <-time.After(defaultTestShortTimeout):
+				isTimeout = true
```

Comment:
The race detector might complain about this. This variable is now written to from two goroutines. An alternative is to use a channel, and have these goroutines do a non-blocking send on the channel whenever they hit the timeout. And instead of checking `if isTimeout { ... }`, you could have to read the channel to see if there are any errors on the channel.

This is an example of a test where we use an error channel: https://github.com/grpc/grpc-go/blob/8628e075df225d916eaeaec1488af4dcc02805ec/test/xds/xds_server_serving_mode_test.go#L113. But unlike that test, you don't have to push a `nil` error at the end of the goroutine. Instead when reading from it, perform a non-blocking read. Something like:

```golang
select {
case <-errCh:
    t.Fatalf("Timeout when expecting the onMessage() callback to be invoked")
default:
}
```